### DISCUSSION TITLE
Step executes but job is skipped for the same condition

### DISCUSSION THREAD

**Discussion by brunokc** [community/community#61792] [Actions] [answered]
### Select Topic Area

Question

### Body

Hi,

I'm new to workflows and actions but after reading the docs and looking through existing workflows, I was able to put something together. 

I have a set of workflows that call each other and the summary of they do is: 

- Run every day on a schedule. Execute the workflows/steps below on each branch (matrix strategy):
  - Check if a version has been updated (via git). If the version has changed:
    - Update the version in the version file (1)
    - Kick off a build (2)
    - Release the new version

Today the workflow ran and found a new version. The step that updates the version file (1) ran and updated the file just fine. The job that kicks off the build was skipped. Both the step and the job check the same condition. Why did the job get skipped?

Today's run: https://github.com/brunokc/pve-kernel-builder/actions/runs/5664091716/job/15346819098

Workflows:

- [trigger-kernel-check.yml](https://github.com/brunokc/pve-kernel-builder/blob/main/.github/workflows/trigger-kernel-check.yml): the schedule and where the matrix strategy is implemented. Call new-pve-kernel-release-check.yml below.
- [new-pve-kernel-release-check.yml](https://github.com/brunokc/pve-kernel-builder/blob/main/.github/workflows/new-pve-kernel-release-check.yml): 
  - Job **check-new-kernel**: checks for a new version and saves the new version if found. See step "Save new kernel version", which corresponds to (1) above.
  - Job **build-kernel**: invokes the build workflow (which corresponds to (2)). See condition, which uses the output of job **check-new-kernel** above (the same condition as (1)).

new-pve-kernel-release-check.yml:

```
name: Check for new PVE kernel release

on:
  [...]
  workflow_call:
    inputs:
      branch:
        description: 'Branch to check for'
        type: string
        required: true
    secrets:
      token:
        description: 'Token used in privileged operations'
        required: false

[...]
jobs:
  check-for-new-kernel:
    name: Check for new kernel release
    runs-on: ubuntu-latest
    #if: github.event.repository.owner.id == github.event.sender.id
    outputs:
      new-kernel-available: ${{ steps.check-version.outputs.status == 'needs-update' }}
      kernel-version: ${{ steps.check-version.outputs.kernel-version }}
      proxmox-version: ${{ steps.check-version.outputs.proxmox-version }}
    permissions:
      contents: write

    steps:
    [...]
    - name: Check for latest kernel version
      id: check-version
      run: |
        config_path='${{ github.workspace }}/config/${{ inputs.branch }}/version'
        cur_abi_ver=""
        if [[ -e ${config_path} ]]; then
            cur_abi_ver=`yq .version.kernel ${config_path}`
        fi
        echo "Current cached kernel version for branch ${{ inputs.branch }}: ${cur_abi_ver}"
        proxmox_ver=`curl -s "https://git.proxmox.com/?p=pve-kernel-meta.git;a=shortlog;h=refs/heads/${{ inputs.branch }}" | \
            grep -oP "bump version to \K[^<]*" | head -n 1`
        url="https://git.proxmox.com/?p=pve-kernel.git;a=shortlog;h=refs/heads/${{ inputs.branch }}"
        payload=`wget $url -qO shortlog.html`
        ver=`grep -oP "bump version to \K[^<,]+" shortlog.html | head -n 1`
        abi_ver=`grep -oP "update ABI file for \K[^<,]+" shortlog.html | head -n 1`
        #kernel_ver="pve-kernel-${abi_ver}-${ver}"
        if [[ ${abi_ver} == ${cur_abi_ver} ]]; then
            echo "Kernel ABI version ${abi_ver} for branch ${{ inputs.branch }} is up to date. Nothing to do."
            echo "status=up-to-date" >> $GITHUB_OUTPUT
        else
            echo "New kernel ABI version avaiable for branch ${{ inputs.branch }}: ${abi_ver}. Will update repository."
            mkdir -p `dirname ${config_path}`
            sudo echo -e "version:\n  proxmox: ${proxmox_ver}\n  kernel: ${abi_ver}" > ${config_path}
            echo "kernel-version=${abi_ver}" >> $GITHUB_OUTPUT
            echo "proxmox-version=${proxmox_ver}" >> $GITHUB_OUTPUT
            echo "status=needs-update" >> $GITHUB_OUTPUT
        fi
        rm -f shortlog.html

    - name: Save new kernel version
      continue-on-error: true
      if: steps.check-version.outputs.status == 'needs-update'
      run: |
        git config --local user.email "github-actions@github.com"
        git config --local user.name "Github Actions"
        git pull
        git add config/**
        git commit -m 'Update version for branch ${{ inputs.branch }} to ${{ steps.check-version.outputs.kernel-version }}'
        git push

  build-kernel:
    name: Build new kernel
    needs: check-for-new-kernel
    if: needs.check-for-new-kernel.outputs.new-kernel-available == 'true'
    uses: ./.github/workflows/build-pve-kernel.yml
    with:
      branch: ${{ inputs.branch }}
[...]
```

If you look at the condition for build-kernel (`needs.check-for-new-kernel.outputs.new-kernel-available == 'true'`), you'll see it resolves to the same one being used by the 'Save new kernel version' step (`steps.check-version.outputs.status == 'needs-update'`) due to this:

```
    outputs:
      new-kernel-available: ${{ steps.check-version.outputs.status == 'needs-update' }}
```

Why didn't the `build-kernel` job run?

**Comment by Akash1134** [comment]
Hi @brunokc ,

It looks like that the `needs.check-for-new-kernel.outputs.new-kernel-available == 'true'` evaluated to `false`. This condition checks the output of the `check-for-new-kernel` job, which sets the `new-kernel-available` output to `true` if a new kernel version is available and needs to be updated.

One possible reason why the `build-kernel `job was skipped is that the `check-for-new-kernel` job did not set the `new-kernel-available` output to true. This could be due to a problem with the `check-for-new-kernel` job itself, such as an issue with the curl or grep commands used to retrieve the kernel version.

Can you try running the `check-for-new-kernel` job manually and checking its output to see if the `new-kernel-available` output is set to `true`? Also, check the logs for the `check-for-new-kernel` job to see if there are any errors or issues that may be causing it to fail.

Thanks!

**Reply by brunokc** [reply]
Hi @Akash1134,

Thanks for taking the time to look into my question. 

I'm pretty sure `check-for-new-kernel` was successful. What tells me that is the fact that the step called `Save new kernel version`, which only runs if `steps.check-version.outputs.status == 'needs-update'`, did run. Since the condition for the `new-kernel-available` job is set as the same expression, it should be true as well, right? What am I missing?

**Comment by brunokc** [comment]
Missed you questions at the end:

> Can you try running the `check-for-new-kernel` job manually and checking its output to see if the `new-kernel-available` output is set to `true`? Also, check the logs for the `check-for-new-kernel` job to see if there are any errors or issues that may be causing it to fail.

I can try. I'm not sure how to check the output values though. Do I have to explicitly print them out? This is all I see at the end of the `check-for-new-kernel` job:

```
2023-07-26T02:49:26.2926569Z Evaluate and set job outputs
2023-07-26T02:49:26.2947842Z Set output 'new-kernel-available'
2023-07-26T02:49:26.2948970Z Set output 'kernel-version'
2023-07-26T02:49:26.2949589Z Set output 'proxmox-version'
2023-07-26T02:49:26.2950393Z Cleaning up orphan processes
```

The full logs are here: https://github.com/brunokc/pve-kernel-builder/actions/runs/5664091716/job/15346819098. I don't see any errors. Everything ran the way they should have.

**Reply by ChristopherHX** [reply]
> Do I have to explicitly print them out?

No, just press rerun and select enable debug logs. It will print more about your outputs.

_Your problem sounds like a Actions Service bug and you may want create a GitHub Support ticket with a reproduction repo of the bug, if your GitHub plan allows that_. There were at least two problems with needs contextdata and reusable workflows in the past, all results are really confusing.

GitHub fixed the usage in the matrix strategy at least for most use cases via a GitHub support case.

**Reply by brunokc** [reply]
Nice, thank you! Great tip! Look at this:

```
##[debug]Starting: Complete job
Evaluate and set job outputs
##[debug]Evaluating: (steps.check-version.outputs.status == 'needs-update')
##[debug]Evaluating Equal:
##[debug]..Evaluating Index:
##[debug]....Evaluating Index:
##[debug]......Evaluating Index:
##[debug]........Evaluating steps:
##[debug]........=> Object
##[debug]........Evaluating String:
##[debug]........=> 'check-version'
##[debug]......=> Object
##[debug]......Evaluating String:
##[debug]......=> 'outputs'
##[debug]....=> Object
##[debug]....Evaluating String:
##[debug]....=> 'status'
##[debug]..=> 'needs-update'
##[debug]..Evaluating String:
##[debug]..=> 'needs-update'
##[debug]=> true
##[debug]Expanded: ('needs-update' == 'needs-update')
##[debug]Result: true
[...]
##[debug]Finishing: Complete job
```

So, if I'm reading this correctly, it looks like we're setting the output for `new-kernel-available` to `true`, but still, the job is being skipped. This does look like a bug in the Actions Service. I use a free account. Do you know if I can still report these kinds of issues? Do you know how to open that ticket?

**Reply by ChristopherHX** [reply]
Sadly we both have the free Plan / account and we must hope they still see this discussion.
> Do you know if I can still report these kinds of issues?

No, this is the right place for free account users. We have no rights for paid plan support.

The last time has been the author of this discussion (https://github.com/orgs/community/discussions/32023) so friendly and opened a support request for a similar bug.

**Reply by ChristopherHX** [reply]
You may also want to try to prefix the condition with `"!cancelled() && !failure() && <your cond>"` to enshure your job is not skipped due to previous skipped jobs. However it doesn't seem your problem.

**Reply by brunokc** [reply]
This is so weird. 

I added a new job called `pre-build-kernel` that prints the value of the condition. It looks like this now:

```
[...]
  pre-build-kernel:
    name: Pre-build new kernel
    runs-on: ubuntu-latest
    needs: check-for-new-kernel
    steps:
      - name: Print new-kernel-available value
        run: |
          echo "New kernel available: ${{ needs.check-for-new-kernel.outputs.new-kernel-available }}"
      
  build-kernel:
    name: Build new kernel
    needs: check-for-new-kernel
    if: needs.check-for-new-kernel.outputs.new-kernel-available == true
    uses: ./.github/workflows/build-pve-kernel.yml
    with:
      branch: ${{ inputs.branch }}
[...]
```

The [output](https://github.com/brunokc/pve-kernel-builder/actions/runs/5675609500/job/15381110241) was this:

```
[...]
2023-07-27T01:51:04.1895696Z Complete job name: Check for new release for each branch (master) / Pre-build new kernel
2023-07-27T01:51:04.3093371Z ##[group]Run echo "New kernel available: true"
2023-07-27T01:51:04.3094016Z [36;1mecho "New kernel available: true"[0m
2023-07-27T01:51:04.3951377Z shell: /usr/bin/bash -e {0}
2023-07-27T01:51:04.3951845Z env:
2023-07-27T01:51:04.3952329Z   DEBIAN_FRONTEND: noninteractive
2023-07-27T01:51:04.3953201Z   TOKEN: ***
2023-07-27T01:51:04.3953692Z ##[endgroup]
2023-07-27T01:51:04.4639328Z New kernel available: true
2023-07-27T01:51:04.4974872Z Cleaning up orphan processes
```

So, the new job prints `${{ needs.check-for-new-kernel.outputs.new-kernel-available }}` as `true`. Even though I'm using the same expression as condition for the actual `build-kernel` job, it's still getting skipped.

I'm wondering now if the syntax of the `if` statement is wrong somehow. I tried `if: needs.check-for-new-kernel.outputs.new-kernel-available == true` and `if: needs.check-for-new-kernel.outputs.new-kernel-available == 'true'` ('true' in quotes). The docs say I don't need to use `${{ }}` with 'if` -- it says it will stringize it. I'll give it a try though. Do you see anything wrong here?

**Reply by ChristopherHX** [reply]
Yes the correct way for your if is true in quotes

> `if: needs.check-for-new-kernel.outputs.new-kernel-available == 'true'` ('true' in quotes)

> Do you see anything wrong here?

- That your recent commits removed the required single quotes
- These simple tests are working: https://github.com/ChristopherHX/newcomposite-sample/actions/runs/5680875072/job/15395846397, https://github.com/ChristopherHX/newcomposite-sample/actions/runs/5680992945

To be honest, due to lack of debug outputs of job if expressions is the Actions Service a big black box.

You can try to test your reusable workflow without matrix. (You are stress testing the GitHub Actions Service, by using matrix + needs in if of an additional reusable workflow depth)

_Maybe try to run a simplified version through this tool https://github.com/ChristopherHX/runner.server, if this had no issue it increase the possibility of a bug._

**Reply by brunokc** [reply]
What a nightmare. 😞 I'm sure part of this is my inexperience with workflows in Github, but this ought to be much easier than this.

@ChristopherHX, thank you for giving me a few options to try. I haven't gone through all of them (haven't had enough time to try them all), but I did try some.

First, I put the quotes back in: `if: needs.check-for-new-kernel.outputs.new-kernel-available == 'true'`

Second, adding `!cancelled && !failed() && <my cond>` at first worked. The job got picked up and ran. However, the job is now running even when `<my cond>` is `false`, which makes no sense. As an example, when I trigger the top level job, it uses the matrix strategy to kick off 2 instances. It should run to completion for one of them, it should get short-circuited for the second. Well, the job with that condition runs for both instances. 🤷‍♂️ Now, to through some randomness in all this, even the job that I expected to run eventually gets cancelled. The last time I tried it, it got cancelled 2h26min into it (it usually takes 3h30m). 😠 

Add to this the friction in getting the inner dev loop going (code, test, fix, repeat), this just makes me want to give up.

I might throw this whole idea of using reusable workflows away (which makes for better code isolation) and put everything in a giant workflow to see if the problems go away.

**Reply by brunokc** [reply]
Well, I found out why my job was getting cancelled. Apparently, I needed to set `fail-fast: false` on the strategy to prevent skips on one instantiation of the matrix to cancel jobs on the other instantiations. Ahead with more testing...

**Comment by brunokc** [comment, marked as answer]
This has finally been solved. An extraneous if condition that escaped inspection (it was in the called workflow, not the caller) was preventing execution. Man, that took way too long to detect, partially due to lack of time to investigate, but partially because of lack of debugging output from GitHub to explain why workflows or jobs get skipped. 

If you have a chance, please upvote https://github.com/orgs/community/discussions/60882 so that debug output can be added to explain why workflows and jobs get skipped.
